Add conditional in require_everything to support arity change in Rails 6+

edwinthinks opened the pull request:
Hey there!

I hope this PR helps :)

### Motivation

I stumbled upon an issue when trying running `srb init` in my application running on Rails 6+ (edge). I encountered a arity issue in which the call to `_all_load_paths` in rails.

Here is the error:
```
Traceback (most recent call last):
	9: from /Users/emak/.rbenv/versions/2.6.0/lib/ruby/gems/2.6.0/gems/sorbet-0.4.4289/bin/srb-rbi:232:in `<main>'
	8: from /Users/emak/.rbenv/versions/2.6.0/lib/ruby/gems/2.6.0/gems/sorbet-0.4.4289/bin/srb-rbi:98:in `main'
	7: from /Users/emak/.rbenv/versions/2.6.0/lib/ruby/gems/2.6.0/gems/sorbet-0.4.4289/bin/srb-rbi:227:in `block in make_step'
	6: from /Users/emak/.rbenv/versions/2.6.0/lib/ruby/gems/2.6.0/gems/sorbet-0.4.4289/lib/gem-generator-tracepoint.rb:38:in `main'
	5: from /Users/emak/.rbenv/versions/2.6.0/lib/ruby/gems/2.6.0/gems/sorbet-0.4.4289/lib/gem-generator-tracepoint/tracer.rb:70:in `trace'
	4: from /Users/emak/.rbenv/versions/2.6.0/lib/ruby/gems/2.6.0/gems/sorbet-0.4.4289/lib/gem-generator-tracepoint.rb:39:in `block in main'
	3: from /Users/emak/.rbenv/versions/2.6.0/lib/ruby/gems/2.6.0/gems/sorbet-0.4.4289/lib/require_everything.rb:20:in `require_everything'
	2: from /Users/emak/.rbenv/versions/2.6.0/lib/ruby/gems/2.6.0/gems/sorbet-0.4.4289/lib/require_everything.rb:44:in `require_all_files'
	1: from /Users/emak/.rbenv/versions/2.6.0/lib/ruby/gems/2.6.0/gems/sorbet-0.4.4289/lib/require_everything.rb:108:in `excluded_rails_files'
/Users/emak/.rbenv/versions/2.6.0/lib/ruby/gems/2.6.0/bundler/gems/rails-6a033a06855f/railties/lib/rails/engine.rb:709:in `_all_load_paths': wrong number of arguments (given 0, expected 1) (ArgumentError)
```

On closer inspect the code of https://github.com/sorbet/sorbet/blob/df6085e6ae9846f033064513ea14f069c68b0bf9/gems/sorbet/lib/require_everything.rb#L108

I've found that it conflicts with the changes made on rails which changes the arity of the `_all_load_paths` method. The related commit is here - https://github.com/rails/rails/commit/b6e17b6a4b67ccc9fac5fe16741c3db720f00959

I hope that this PR can help individuals choosing to Rails 6+ whilst support those using previous versions of rails. The solution I propose is to check the arity of the method and call the method appropriately depending on the arity check

### Test plan

I didn't see any existing files for tests for files under this directory. 

See included automated tests.

edwinthinks commented:
@ptarjan hey there! Sure I can look into adding a test for this change. Using something other reflection sounds better :). Hope to have a update in soon on this PR.

edwinthinks commented:
Hey @ptarjan, I've updated the PR to use `Gem::Version` instead of reflection. I'am having difficulty figuring how to add tests in the desired format. Could I get some hints or suggestions here? Thanks!

ptarjan commented:
Copy this test https://github.com/sorbet/sorbet/tree/master/gems/sorbet/test/snapshot/partial/codecov/src but name it https://github.com/sorbet/sorbet/tree/master/gems/sorbet/test/snapshot/partial/rails6/src and setup the `Gemfile` to test that if we have Rails 6 in it completes.  This is the script you run to check the test works https://github.com/sorbet/sorbet/blob/master/gems/sorbet/test/snapshot/driver.sh

edwinthinks commented:
@ptarjan hey there a bit of update on my endeavor. I ran into some issues trying to get rails6 in the tests. Here are some concerns.

- Rails 6 requires Ruby 2.5 and causes the build to fail with version mismatch
```
[ .. ] Running test:  gems/sorbet/test/snapshot/test_one.sh gems/sorbet/test/snapshot/partial/rails6  --verbose
--
  | [ .. ] ├─ PWD:       /usr/local/var/buildkite-agent/builds/desk-cloud/sorbet/sorbet
  | [ .. ] ├─ test_dir:  /usr/local/var/buildkite-agent/builds/desk-cloud/sorbet/sorbet/gems/sorbet/test/snapshot/partial/rails6
  | [ .. ] ├─ actual:    /var/folders/xt/yvc1sy1n2gj05m5r2n2d7c280000gn/T/tmp.DiX4BTdx
  | Fetching gem metadata from https://rubygems.org/............
  | zeitwerk-2.1.6 requires ruby version >= 2.4.4, which is incompatible with the
  | current version, ruby 2.4.3p205
  | [ERR!] └─ 'bundle install' failed. See output above.
```

- Wondering if it is too soon to have this PR since the latest rails version is at `6.0.0.rc1` which does not require this sort of change yet. This is expected to relevant as soon as they release `6.0.0` though.

Thoughts? Thanks :)

manhhung741 commented:
I believe this is becoming relevant now because the change has been released to Rails 6.0.0.rc.
I got this error when running sorbet on the newest release of Rails 6.0.0.rc

```
"/home/dev/.rbenv/versions/2.5.5/lib/ruby/gems/2.5.0/gems/railties-6.0.0.rc2/lib/rails/engine.rb:711:in `_all_load_paths': wrong number of arguments (given 0, expected 1) (ArgumentError)\n" +
"\tfrom /home/dev/.rbenv/versions/2.5.5/lib/ruby/gems/2.5.0/gems/sorbet-0.4.4496/lib/require_everything.rb:114:in `excluded_rails_files'\n" +
"\tfrom /home/dev/.rbenv/versions/2.5.5/lib/ruby/gems/2.5.0/gems/sorbet-0.4.4496/lib/require_everything.rb:44:in `require_all_files'\n" +
"\tfrom /home/dev/.rbenv/versions/2.5.5/lib/ruby/gems/2.5.0/gems/sorbet-0.4.4496/lib/require_everything.rb:20:in `require_everything'\n" +
"\tfrom /home/dev/.rbenv/versions/2.5.5/lib/ruby/gems/2.5.0/gems/sorbet-0.4.4496/lib/hidden-definition-finder.rb:64:in `require_everything'\n" +
"\tfrom /home/dev/.rbenv/versions/2.5.5/lib/ruby/gems/2.5.0/gems/sorbet-0.4.4496/lib/hidden-definition-finder.rb:43:in `main'\n" +
"\tfrom /home/dev/.rbenv/versions/2.5.5/lib/ruby/gems/2.5.0/gems/sorbet-0.4.4496/lib/hidden-definition-finder.rb:38:in `main'\n" +
"\tfrom /home/dev/.rbenv/versions/2.5.5/lib/ruby/gems/2.5.0/gems/sorbet-0.4.4496/bin/srb-rbi:229:in `block in make_step'\n" +
"\tfrom /home/dev/.rbenv/versions/2.5.5/lib/ruby/gems/2.5.0/gems/sorbet-0.4.4496/bin/srb-rbi:221:in `main'\n" +
"\tfrom /home/dev/.rbenv/versions/2.5.5/lib/ruby/gems/2.5.0/gems/sorbet-0.4.4496/bin/srb-rbi:234:in `<main>'\n"
```

connorshea commented:
Yup, this is blocking me from upgrading Rails now that 6.0.0.rc2 includes the change to this method.

edwinthinks commented:
@connorshea hey there, thanks for the comment. I've thought about being able to load an older version of rails and override the version number of rails for the test. However, I haven't had any luck getting it to work. Thoughts on wither this a good route or not?

connorshea commented:
@edwinthinks wouldn’t that not work because if we spoofed the version as being 6.0 when it was actually 5.2.3, Rails would still have the method defined without the parameter? I think this should probably just be merged without a test unless the Sorbet team has a better idea on how to resolve the ruby version problem.

manhhung741 commented:
Can we allow this particular test to fail for Ruby 2.5 and less? Or skip the test when Rails version is 6.0 & ruby version < 2.5. In sorbet rails, we configure the test matrix so that it doesn't run for those combinations

jez commented:
@elliottt handing this off to you after discussion this morning about Ruby 2.6 / testing / etc.

elliottt commented:
Could you rebase this off of master, and add your test to the exclude set defined in [gems/sorbet/test/snapshot/BUILD](https://github.com/sorbet/sorbet/blob/master/gems/sorbet/test/snapshot/BUILD#L34)? I'm working separately on support for snapshot testing with ruby-2.6, but I don't want that work to gate this PR being merged.

jasnow commented:
Can you upgrade to Rails 6.0.0.rc2 while you are at it? Thanks.

edwinthinks commented:
@elliottt thanks! I've rebased this PR off of a master and added it to the exclusion list as you mentioned.

@jasnow I've also upgraded the version to 6.0.0rc2 and changed the conditional to:
```
Gem::Version.new(rails.version) >= Gem::Version.new('6.0.0.rc2')
```
^ I think this makes it a little clearer when this starts becoming an issue.

elliottt commented on gems/sorbet/test/snapshot/partial/rails6/src/.ruby-version:
```diff
@@ -0,0 +1 @@
+2.5.0
```
I think this file can be removed.

elliottt commented:
Could you remove the `//gems/sorbet/test/snapshot/partial/rails6/src/.ruby-version`, and commit the results of running `./tools/scripts/format_build_files.sh`?

edwinthinks commented:
@elliottt thanks! updated the PR based on the feedback :)
